miden-crypto/src/merkle/smt/partial.rs

```diff
@@ -176,7 +176,19 @@ impl PartialSmt {
         //   PartialSmt::insert, this will not error for such empty leaves whose merkle path was
         //   added, but will error for otherwise non-existent leaves whose paths were not added,
         //   which is what we want.
+        let prev_entries = self
+            .0
+            .leaves
+            .get(&current_index.value())
+            .map(|leaf| leaf.num_entries())
+            .unwrap_or(0);
+        let current_entries = leaf.num_entries();
         self.0.leaves.insert(current_index.value(), leaf);
+        if current_entries > prev_entries {
+            self.0.num_entries += (current_entries - prev_entries) as usize;
+        } else {
+            self.0.num_entries -= (prev_entries - current_entries) as usize;
+        }
```

Comment:
It's a bit unfortunate that we have to do it this way, but I don't have a better suggestion. I think it would be good to add a small test to specifically test `PartialSmt::num_entries`, just to make sure this is covered in case of future refactorings.